lib/shared/src/event/log_event/mod.rs

```diff
@@ -0,0 +1,831 @@
+#![allow(clippy::needless_collect)]
+
+pub mod lua;
+#[cfg(test)]
+mod test;
+
+use crate::{event::*, lookup::*};
+use serde::{Deserialize, Serialize};
+use std::str::FromStr;
+use std::{
+    collections::{btree_map::Entry, BTreeMap, HashMap},
+    convert::{TryFrom, TryInto},
+    fmt::Debug,
+    iter::FromIterator,
+};
+use tracing::{debug, error, instrument, trace, trace_span};
+
+/// A map of [`crate::event::Value`].
+///
+/// The inside of an [`Event::Log`](crate::event::Event) variant of [`crate::event::Event`].
+///
+/// This type supports being interacted with like a regular old
+/// [`BTreeMap`](std::collections::BTreeMap), or with special (unowned) [`crate::event::Lookup`] and
+/// (owned) [`crate::event::LookupBuf`] types.
+///
+/// Transparently, as a normal [`BTreeMap`](std::collections::BTreeMap):
+///
+/// ```rust
+/// use shared::{event::*, lookup::*};
+/// let mut event = LogEvent::default();
+/// event.insert(String::from("foo"), 1);
+/// assert!(event.contains("foo"));
+/// assert_eq!(event.get("foo"), Some(&Value::from(1)));
+/// ```
+///
+/// Using remap-style lookups:
+///
+/// ```rust
+/// use shared::{event::*, lookup::*};
+/// let mut event = LogEvent::default();
+/// let lookup = LookupBuf::from_str("foo[0].(bar | bat)").unwrap();
+/// event.insert(lookup.clone(), 1);
+/// assert!(event.contains(&lookup));
+/// assert_eq!(event.get(&lookup), Some(&Value::from(1)));
+/// ```
+///
+/// It's possible to access the inner [`BTreeMap`](std::collections::BTreeMap):
+///
+/// ```rust
+/// use shared::{event::*, lookup::*};
+/// let mut event = LogEvent::default();
+/// event.insert(String::from("foo"), 1);
+///
+/// use std::collections::BTreeMap;
+/// let _inner: &BTreeMap<_, _> = event.inner();
+/// let _inner: &mut BTreeMap<_, _> = event.inner_mut();
+/// let inner: BTreeMap<_, _> = event.take();
+///
+/// let event = LogEvent::from(inner);
+/// ```
+///
+/// There exists a `log_event` macro you may also utilize to create this type:
+///
+/// ```rust
+/// use shared::{log_event, event::*, lookup::*};
+/// let event = log_event! {
+///     "foo" => 1,
+///     LookupBuf::from_str("bar.baz").unwrap() => 2,
+/// }.into_log();
+/// assert!(event.contains("foo"));
+/// assert!(event.contains(Lookup::from_str("foo").unwrap()));
+/// ```
+#[derive(PartialEq, Debug, Clone, Default, Serialize, Deserialize)]
+pub struct LogEvent {
+    #[serde(flatten)]
+    fields: BTreeMap<String, Value>,
+}
+
+impl LogEvent {
```

Comment:
Given that `LogEvent` is essentially a wrapper over `BTreeMap` and `Value::Map` is also a wrapper over `BTreeMap`, is there any reason why we can't share more lookup code between this and `Value`?

Reply:
That could absolutely be done! I actually did this at one point along my refactoring. 
Here's why I didn't:

* `LogEvent`, as Vector currently works, **must** always be a map, making `LogEvent` a `Value` would mean we can't guarantee that `LogEvent` is indeed a value. We'd have to use things like `debug_assert!()` and `Value::is_map()` quite often. It felt kind of unsafe.
* At various states the code did in fact differ, so they were differing to allow that. I should audit to see if I can combine them now.

I agree with you that this is a worthwhile exploration to take, and I'll have a try at it.

Reply:
Yeah, I think making `LogEvent` a `Value` would likely be a mistake for the reasons you outlined.  I'm thinking more things like:

```rust
  impl LogEvent {
    pub fn get<'a>(&self, lookup: impl Into<Lookup<'a>> + Debug) -> Option<&Value> {
       self.as_value().get(lookup)
    }
 
  ....
```

And then `as_value` just wraps the `LogEvent::fields` in a `Value::Map`.

I realise there are probably 100's of gotchas to this I haven't thought through..

Reply:
In order to do this, `LogEvent` would have to store a `Value` itself, or clone the BtreeMap and transmute it into a value each time (which it then would not be able to return since it can't return a borrow to a value owned by the current scope). The prior is kind of problematic for the reasons outlined, the former is hecking expensive.